### ISSUE TITLE
[Question] Why SAETrainer do not normalize W_dec?

### ISSUE THREAD

**Issue by jasonrichdarmawan** [jbloomAus/SAELens#544]
# Questions

Assumption from the code:
1. `mse_loss` depends on `W_enc` and `W_dec`.
2. `l1_loss` depends on `W_enc` (and `W_dec` norm).

If we don't normalize `W_dec`, then wouldn't it be possible to make `W_enc` very small to make `l1_loss` small and adjust `W_dec` accordingly to make `mse_loss` constant?

In other words, the model cheats to lower the `l1_loss`

No code indicate normalizing `W_dec` in `_train_step` function

https://github.com/jbloomAus/SAELens/blob/3432f0059bc1d90119fe48f88e17ac07aea3a620/sae_lens/training/sae_trainer.py#L230-L281

`training_forward_pass` function

https://github.com/jbloomAus/SAELens/blob/3432f0059bc1d90119fe48f88e17ac07aea3a620/sae_lens/saes/sae.py#L917-L963

`calculate_aux_loss` function

https://github.com/jbloomAus/SAELens/blob/3432f0059bc1d90119fe48f88e17ac07aea3a620/sae_lens/saes/gated_sae.py#L173-L202

**Comment by chanind** [comment]
Thank you for finding this! It does seem like we should scale the pi_gate acts by the decoder norm, like we do in StandardSAE: https://github.com/jbloomAus/SAELens/blob/3432f0059bc1d90119fe48f88e17ac07aea3a620/sae_lens/saes/standard_sae.py#L148. This isn't necessary for Topk/BatchTopk and JumpReLU since there's no L1 penalty, which is why it's not in the base SAE class or in the trainer.

My guess is that this broke in the v6 transition, since it used to be possible to normalize the decoder on each train step, but this was removed in v6 as this is no longer considered good practice. Likely the GatedSAE implementation dates to a time when it was standard practice to forcibly normalize the decoder on each train step. I'll make a PR to fix this and add unit tests to avoid this regression in the future.

**Comment by chanind** [comment]
Wait looking at this more we are correctly scaling by decoder norm in GatedSAEs here: https://github.com/jbloomAus/SAELens/blob/3432f0059bc1d90119fe48f88e17ac07aea3a620/sae_lens/saes/gated_sae.py#L192

**Comment by chanind** [comment]
Are you actually seeing the decoder norm getting very small when you train or are you just speculating? It looks like things are being handled correctly in the code as far as I can tell.

**Comment by jasonrichdarmawan** [comment]
Ah, sorry, I guess you can scale the `pi_gate_act` with `W_dec.norm(dim=-1)` to prevent the model from cheating. But wouldn't that make the `L1 loss` flow to `W_dec`? 

Which we don't want that to happen?

The same argument against why we should use `via_gate_reconstruction = pi_gate_act @ self.W_dec.detach() + self.b_dec.detach()`

Would you mind to share your thoughts about this?

[WandB](https://wandb.ai/jasonrichdarma/sonar_sae?nw=nwuserjasonrichdarmawan)

**Comment by jasonrichdarmawan** [comment]
> Are you actually seeing the decoder norm getting very small when you train or are you just speculating? It looks like things are being handled correctly in the code as far as I can tell.

I didn't log the decoder norm and just speculating. I am sorry for taking your attention on this matter

For context: I was looking for why `via_gate_reconstruction = pi_gate_act @ self.W_dec.detach() + self.b_dec.detach().detach()` cause the training to fail to learn anything useful

**Comment by chanind** [comment]
Apologies for the trouble with the Gated SAEs, and thank you for reporting your experiences with this! Does this PR fix things? https://github.com/jbloomAus/SAELens/pull/545

I think you're correct that we should be detaching `W_dec` and `b_dec`. I suspect the reason your previous attempt didn't work is that it didn't also remove the decoder norm scaling where things were multiplied by `self.W_dec.norm(dim=1)`

**Comment by jasonrichdarmawan** [comment]
I believe we need to normalize `W_dec`. I assume PR #545 allows the model to cheat L1

**Comment by jasonrichdarmawan** [comment]
Hi, 

Should we use normalized decoder weights to compute $$L_\text{reconstruct}$$ to prevent $$L_1$$ cheating by decreasing the encoder weights and increasing the decoder weights?

For context: I am not sure whether MSE loss have such incentive to do so when there is $$L_1$$ loss. Please let me know if my assumption is flawed (ps: I do feel it's flawed. I want to learn)

From my understanding, `dictionary_learning` use `ConstrainedAdam` which normalize decoder weights automatically (not sure how it knows which is decoder and encoder). The [code](https://github.com/saprmarks/dictionary_learning/blob/60ec6bf5264944d64a4ca271f45a29ebfb9d4946/dictionary_learning/trainers/gdm.py#L61-L66)

Without normalized decoder weights when computing the $$L_{reconstruct}$$, the norm is above 1.

<img width="1792" height="1137" alt="Image" src="https://github.com/user-attachments/assets/08b21480-302d-47d0-8e9d-41380748dffb" />

https://github.com/jbloomAus/SAELens/blob/47938a0cbc53f689b3cab75779a3801129f34b21/sae_lens/saes/sae.py#L922-L927

https://github.com/jbloomAus/SAELens/blob/47938a0cbc53f689b3cab75779a3801129f34b21/sae_lens/saes/gated_sae.py#L75-L87

<img width="1030" height="1106" alt="Image" src="https://github.com/user-attachments/assets/af250cb7-79b4-4f21-8a43-ca20169d210e" />
